server.js

```diff
@@ -2,37 +2,128 @@ import express from 'express'
+    // Clears database
+    await Book.deleteMany()
+
+    // Saves all books from booksData to the database
+    await booksData.forEach(book => new Book(book).save())
+  }
+  seedDatabase()
+}
+
 const port = process.env.PORT || 8080
 const app = express()
 
-// Add middlewares to enable cors and json body parsing
+const listEndpoints = require('express-list-endpoints')
+
+// Middlewares to enable cors and json body parsing, and handling if API service is unavailable
 app.use(cors())
 app.use(bodyParser.json())
+app.use((req, res, next) => {
+  if (mongoose.connection.readyState === 1) {
+    next()
+  } else {
+    res.status(503).json({ error: 'Service unavailable' })
+  }
+})
 
-// Start defining your routes here
+// Root endpoint
 app.get('/', (req, res) => {
-  res.send('Hello world')
+  res.send(listEndpoints(app))
+})
+
+// Route for all books
+app.get('/books', async (req, res) => {
+  const { page, sort, title, author } = req.query
+  const pageNo = +page || 1
+  const perPage = 10
+  // skip: E.g. page 3: 10 * (3-1) = 20, sends 20 as parameter to .skip()
+  // skips index 0-19 so that page 3 starts with the book that has index 20
+  const skip = perPage * (pageNo - 1)
+  const error404 = 'No books found, please try a different query'
+
+  // Used to sort bookslist based on sort query
+  const sortQuery = (sort) => {
+    if (sort === 'rating_dsc') {
+      return { average_rating: -1 }
+    } else if (sort === 'rating_asc') {
+      return { average_rating: 1 }
+    } else if (sort === 'pages_dsc') {
+      return { num_pages: -1 }
+    } else if (sort === 'pages_asc') {
+      return { num_pages: 1 }
+    } else if (sort === 'alphabetical') {
+      return { title: 1 }
+    }
+  }
+
+  // Bookslist with filters based on title and/or authors query
+  // Paginated using limit and skip, change page using page query
+  const booksList = await Book.find({
+    title: new RegExp(title, 'i'),
+    authors: new RegExp(author, 'i')
+  })
+    .sort(sortQuery(sort))
+    .limit(perPage)
+    .skip(skip)
+
+  // Used to display total books, checking booksList length without pagination
+  const booksListNoPagination = await Book.find({
+    title: new RegExp(title, 'i'),
+    authors: new RegExp(author, 'i')
+  })
+  const numBooks = booksListNoPagination.length
+
+  // JSON returned depending on the booksList being empty or not
+  if (booksList.length === 0) {
+    res.status(404).json({ error: error404 })
+  } else {
+    res.json({
+      total_books: numBooks,
+      total_pages: Math.ceil(numBooks / perPage),
+      page: pageNo,
+      books: booksList
+    })
+  }
+})
+
+// Route for single book using ISBN-13 as param
+app.get('/books/:isbn13', async (req, res) => {
+  const { isbn13 } = req.params
+  const error404 = `No book found with ISBN-13 ${isbn13}`
+  const error400 = `${isbn13} is not a valid ISBN-13`
+  // Used to check if the isbn13 param starts with 978 followed by 10 digits
+  const isbnCheck = /^(978)([0-9]{10})$/
```

Comment:
Nice usage of regular expression!

Reply:
I was happy when it worked out, so thanks!

And thank you for the fastest review in history 🎉

Reply:
You're welcome :)